pageserver: fixes for creation operations overlapping with shutdown/startup

jcsp opened the pull request:
## Problem

For #6423, creating a reproducer turned out to be very easy, as an extension to test_ondemand_activation.

However, before I had diagnosed the issue, I was starting with a more brute force approach of running creation API calls in the background while restarting a pageserver, and that shows up a bunch of other interesting issues.



In this PR:
- Add the reproducer for #6423 by extending `test_ondemand_activation` (confirmed that this test fails if I revert the fix from https://github.com/neondatabase/neon/pull/6430)
- In timeline creation, return 503 responses when we get an error and the tenant's cancellation token is set: this covers the cases where we get an anyhow::Error from something during timeline creation as a result of shutdown.
- While waiting for tenants to become active during creation, don't .map_err() the result to a 500: instead let the `From` impl map the result to something appropriate (this includes mapping shutdown to 503)
- During tenant creation, we were calling `Tenant::load_local` because no Preload object is provided.  This is usually harmless because the tenant dir is empty, but if there are some half-created timelines in there, bad things can happen.  Propagate the SpawnMode into Tenant::attach, so that it can properly skip _any_ attempt to load timelines if creating.
- When we call upsert_location, there's a SpawnMode that tells us whether to load from remote storage or not.  But if the operation is a retry and we already have the tenant, it is not correct to skip loading from remote storage: there might be a timeline there.  This isn't strictly a correctness issue as long as the caller behaves correctly (does not assume that any timelines are persistent until the creation is acked), but it's a more defensive position.
- If we shut down while the task in Tenant::attach is running, it can end up spawning rogue tasks.  Fix this by holding a GateGuard through here, and in upsert_location shutting down a tenant after calling tenant_spawn if we can't insert it into tenants_map.  This fixes the expected behavior that after shutdown_all_tenants returns, no tenant tasks are running.
- Add `test_create_churn_during_restart`, which runs tenant & timeline creations across pageserver restarts.
- Update a couple of tests that covered cancellation, to reflect the cleaner errors we now return.

## Summary of changes

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

jcsp commented:
Latest failure of test_tenant_delete_races_timeline_creation was due to https://github.com/neondatabase/neon/pull/6387.  Should re-run-this after that merges.

problame reviewed (approved):
`reject generations that go backward`.

It bothers me that we had to implement this check in two places.
Not this PR's fault, it's just exposing a flaw.

---


```test_create_churn_during_restart``` ; how about making this time-based instead of `for _ in range(0, 10)`. I.e., do iterations until  30s have passed.

---

```avoid leaking Tenant``` :  proposal for additional refactor: if `slot_guard.upsert(new_slot)` fails, it should return the `new_slot` that was passed in. Avoids the tenant cloning in the lines above.

---

`carry a GateGuard`

```
// The attach task will carry a GateGuard, so that shutdown() reliably waits for it to drop out if
        // we shut down while attaching.
        let Ok(attach_gate_guard) = tenant.gate.enter() else {
```

how can we ever hit `else`, we just created that tenant object, it's not reachable from anywhere.

---

`tests: update location_conf_churn`: instead of doing `last_state_ps[1] + 1` , we should insead get the generation number from tenant status (I added that to the API output recently). No guarantee that generation numbers are sequential. 

---


`log freeze and flush errors at info!`: strongly opposed to how it's done here. we should only log at info! level if it's during shutdown. Should have `is_cancelled()` / `is_shutdown()` method on the error type. If it's an anyhow::Error, I say we punt on this change for now and refactor `freeze_and_flush` first so it returns an error type that lets us differentiate cancellation from other error causes. 

Also, consider using `report_compact_sources` , see  recent discussion https://neondb.slack.com/archives/C033RQ5SPDH/p1706105202934999

---

`avoid giving 404 for tenants during shutdown`: the comment that you added in `tenant_map_peek_slot` doesn't mention this is so that we don't return 404. Maybe it should? 

---

`allow list for #6385` : you put a TODO there, but, #6385 is closed.

---

`test: update .*`: are these fixup commits to one of your earlier commits? If so, please `git revise` it. 

---

Despite my snarky comments above, great change set!

After you potentially clean up the history using `git revise` or similar, let's pease **merge this as a merge commit**; squashing it will erase too much history, the changes seem related but freestanding.

arpad-m commented on test_runner/regress/test_tenants.py:
```diff
@@ -341,3 +345,71 @@ def not_attaching():
+    env.pageserver.allowed_errors.append(".*Dropped remote consistent LSN updates for tenant .+")
+
+    # TODO: https://github.com/neondatabase/neon/issues/6385
+    env.pageserver.allowed_errors.append(".*Failed to remove temporary initdb directory.*")
```
this is not needed any more since my PR #6451 has been merged, right?

jcsp replied:
Right, I'm removing this commit.

jcsp commented:
>tests: update location_conf_churn: instead of doing last_state_ps[1] + 1 , we should insead get the generation number from tenant status (I added that to the API output recently). No guarantee that generation numbers are sequential.

That was my first instinct, but it's awkward: during tenant shutdown, flush_and_shutdown gets called before Timeline::shutdown is called, so the timeline doesn't really know it is shutting down  :-/ 


> reject generations that go backward.
> 
> It bothers me that we had to implement this check in two places.
> Not this PR's fault, it's just exposing a flaw.

Yep.  If we ultimately wanted to get this down to one location, it would require changing startup process to wait to be `/location_conf`'d (plausible change, in the mid-term).

> test_create_churn_during_restart ; how about making this time-based instead of for _ in range(0, 10). I.e., do iterations until 30s have passed.

I'm kind of ambivalent.  Generally prefer a fixed number of iterations to something time based, so that we're not e.g. doing fewer iterations in debug mode.

> avoid leaking Tenant : proposal for additional refactor: if slot_guard.upsert(new_slot) fails, it should return the new_slot that was passed in. Avoids the tenant cloning in the lines above.

The cloning was already there (it's used to return the Tenant from the function).

Originally I wanted to keep the error simpler, but later I added the distinct ShuttingDown case to carry the Completion: so yeah, it makes sense to carry the slot here too.  

> carry a GateGuard
> 
> // The attach task will carry a GateGuard, so that shutdown() reliably waits for it to drop out if
>         // we shut down while attaching.
>         let Ok(attach_gate_guard) = tenant.gate.enter() else {
> how can we ever hit else, we just created that tenant object, it's not reachable from anywhere.

That's true -- I was just handling error out of habit.  Replaced this with unreachable!();

> tests: update location_conf_churn: instead of doing last_state_ps[1] + 1 , we should insead get the generation number from tenant status (I added that to the API output recently). No guarantee that generation numbers are sequential.

Hmm, the only way the generation numbers jump by more than one in this context is if something goes wrong in /re-attach and the pageserver calls it again.

I've updated this to still just add 1, but also assert that the generation matches from the pageserver API: I don't want to just silently tolerate anything weird that's jumping the generation by more than one, but it's right that we also shouldn't assume the test code gets this right.

> log freeze and flush errors at info!: strongly opposed to how it's done here. we should only log at info! level if it's during shutdown. Should have is_cancelled() / is_shutdown() method on the error type. If it's an anyhow::Error, I say we punt on this change for now and refactor freeze_and_flush first so it returns an error type that lets us differentiate cancellation from other error causes.
> 
> Also, consider using report_compact_sources , see recent discussion https://neondb.slack.com/archives/C033RQ5SPDH/p1706105202934999

I'm replacing this with a tolerance of the log in the test.

> avoid giving 404 for tenants during shutdown: the comment that you added in tenant_map_peek_slot doesn't mention this is so that we don't return 404. Maybe it should?

Added a sentence.

> allow list for #6385 : you put a TODO there, but, https://github.com/neondatabase/neon/issues/6385 is closed.

This is gone now

> test: update .*: are these fixup commits to one of your earlier commits? If so, please git revise it.

Not sure which exact commits drove these without going back and re-testing each commit.

jcsp commented:
Last failures were "still waiting, taking longer than expected..." logs, which makes sense.  Since https://github.com/neondatabase/neon/pull/6451, we will sometimes wait longer for an initdb to finish (and that is what's happening in the logs from these failures).  Added a line to the test to tolerate that.

arpad-m commented:
Yeah in #6451 I had to add allows for still waiting as well.
